File: merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    static bool heapComp(ListNode* a, ListNode* b) {
        return a->val > b->val;
}
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        ListNode head(0);
        ListNode *curNode  = &head;
        vector<ListNode*> v;
        
        for (auto i: lists){
            if (i) {
                v.push_back(i);
            }
        }
        make_heap(v.begin(), v.end(), heapComp);

        while(v.size() > 0){
            curNode->next = v.front();
            pop_heap(v.begin(), v.end(), heapComp);
            v.pop_back();
            curNode = curNode->next;
            if (curNode->next){
                v.push_back(curNode->next);
                make_heap(v.begin(), v.end(), heapComp);
                
            
            }
        }
        return head.next;
    }
};

File: merge-k-sorted-lists/merge-k-sorted-lists.py
class Solution(object):
    def mergeKLists(self, lists):
        if not lists or len(lists) == 0:
            return None
        def mergeTwoLists(l1, l2):
            res = []
            dummy = ListNode()
            l3 = dummy
            while l1 and l2:
                if l1.val > l2.val:
                    l3.next = l2
                    l2 = l2.next
                else:
                    l3.next = l1
                    l1 = l1.next
                l3 = l3.next
            if l1:
                l3.next = l1
            elif l2:
                l3.next = l2
            return dummy.next
        
        while len(lists) > 1:
            l1 = lists.pop()
            l2 = lists.pop()  
            lists.append(mergeTwoLists(l1, l2))
        return lists[0]